### PR TITLE
BSA errata 2: expose APIs under cudnn.jax and cudnn.torch

### PR DESCRIPTION
**Pull request by hwanseoc**
## Summary

Second BSA errata, following #1081. Expose BSA through framework namespaces: move the existing JAX functions to `cudnn.jax` and add lazy aliases of the existing Torch functions under `cudnn.torch`. The three former `_jax` exports through `cudnn` and `BSA` are removed.

```python
import jax
from cudnn.jax import block_sparse_attention as attention
from cudnn.jax import block_sparse_attention_forward, block_sparse_attention_backward

output = jax.jit(lambda q, k, v, i: attention(q, k, v, i, 2))(q, k, v, indices)
dq = jax.grad(lambda q: attention(q, k, v, indices, 2).sum())(q)
```

```python
from cudnn.torch import block_sparse_attention_forward, block_sparse_attention_backward

o, lse = block_sparse_attention_forward(q, k, v, indices, 2)
dq, dk, dv = block_sparse_attention_backward(do, q, k, v, o, lse, indices, 2)
```

`cudnn.torch.block_sparse_attention_fp8_forward` also aliases the existing FP8 forward function. The aliases are the same Python function objects as the existing `cudnn` and `cudnn.BSA` exports, preserving signatures and return values. Torch backward remains explicit; this does not add Torch autograd.

Both namespaces export the existing functions lazily. There is no framework-dispatch layer or new execution wrapper. The Torch API file is identical to develop; the JAX implementation differs only in its three function names. Kernels, validation, XLA stream ordering, workspaces, caches, and custom VJP are unchanged.

## Why

The namespace identifies the framework, so JAX functions do not need `_jax` suffixes. This also keeps the differentiation helper clearly JAX-specific. The newly introduced names have no customer compatibility requirement. Related to #976 and #1081.

## Affected area

Python API exports, documentation, and tests.

## API and compatibility impact

- `cudnn.jax.block_sparse_attention`: output with first-order reverse-mode Q/K/V derivatives.
- `cudnn.jax.block_sparse_attention_forward`: explicit output and LSE.
- `cudnn.jax.block_sparse_attention_backward`: explicit dQ/dK/dV.
- All three accept JAX arrays eagerly or under `jax.jit`, without requiring Torch.
- Remove the former `_jax` exports without aliases. Existing Torch imports and signatures remain unchanged.
- Existing JAX support limits remain: one visible SM100, BF16, blk128, BHSD/BSHD, equal head counts, D=64/128. Torch-only launch options and caller-provided output buffers are not JAX parameters.
- Add `cudnn.torch.block_sparse_attention_forward`, `block_sparse_attention_backward`, and `block_sparse_attention_fp8_forward` as aliases. Existing Torch paths stay supported. Update the API index and examples for both namespaces; other operation families are outside this PR.

## Testing

After merging develop with #1079, the shared namespace handlers preserve both BSA and grouped-GEMM exports. A fresh SM100 container run of `fe_api/jax`, `fe_api/bsa/test_BSA_jax_parity.py`, and `fe_api/grouped_gemm/test_grouped_gemm_canonical_jax.py` with `--require-sm100` passed **110 tests, zero skips** in 173.50 seconds. Current Python sources were mounted over the existing CI native extension; CuTeDSL 4.7.1, JAX 0.11.1, Torch 2.13.0a0, cuDNN 9.24. Changed namespace files passed pre-commit.

Container checks on SM100 with the current Python package sources and tests mounted over a CI-built native extension (frontend 1.30.0). This was not a fresh native build.

- **55 passed, 23 skipped**, Torch/parity/adapter suite after adding the Torch aliases:
  `python -m pytest fe_api/bsa fe_api/test_tensor_adapter.py -q -p no:cacheprovider`
  Torch 2.13.0a0, JAX 0.11.1, CuTeDSL 4.7.1, cuDNN 9.24. Covers Torch forward/backward/FP8, both JAX parity layouts, mixed-framework rejection, lazy Torch alias resolution, identity with the existing exports, and forward/backward through `cudnn.torch` with JAX unavailable. The alias test was observed failing before the exports were added.
- **8 passed, 1 failed**, CPU API-index checks using the complete current Python package sources:
  `python -m unittest discover -s test/python/api_index -p test_api_index.py -v`
  The only runtime-index difference is the existing unrelated `cudnn.sdpa.fwd.SdpaFwdDslSm100.template_params` entry. The new Torch exports match the updated baseline. This same SDPA mismatch appeared in the previous OSS pipeline; the full API-index check is not green.
- Earlier validation of the unchanged JAX implementation: **42 passed** in the torch-free JAX container (JAX 0.10.2.dev20260630+3757395a28, CuTeDSL 4.7.0, cuDNN 9.24):
  `python -m pytest /review/test/python/fe_api/jax --confcutdir=/review/test/python/fe_api/jax -q --require-sm100 -p no:cacheprovider`
  Covers eager/JIT numerics, gradients, asynchronous calls, warmed execution, lazy exports, removed names, and import isolation.
- Changed-file pre-commit and `git diff --check` passed. Performance was not remeasured; these are function aliases, with no added execution wrapper.

## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran pre-commit for changed files and committed formatting changes.
- [x] I reviewed applicable AGENTS.md hard rules.
- [x] I added category, area, and origin labels.

@coderabbitai ignore

### CODE REVIEW

**Comment by hwanseoc** [comment]
@cudnn-ci-bot run oss

**Comment by hwanseoc** [comment]
@cudnn-ci-bot run oss

**Comment by hwanseoc** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `49c506d65f01` against base `adb963c01b6f`.

The final namespace-only change makes sense. The three JAX APIs are exported lazily from cudnn.jax, while the existing Torch API and execution paths remain unchanged. The intentional removal of the new _jax names is documented, and the examples and API index follow the new namespace. No actionable finding.

Validation: Reviewed the complete final diff and the changes since the two earlier reviewed revisions. Torch api.py is byte-identical to merge base b5ff076fb; an AST comparison confirms JAX jax_api.py differs only in its three public function names. Kernels, custom VJP, stream/workspace handling, and cache implementations are unchanged. Exact head 49c506d, SM100 (68 SMs), cuDNN 9.25.1, DSL 4.8, Torch 2.13 and JAX 0.11.0: the JAX suite passed 42/42, covering eager/JIT execution, gradients, asynchronous reuse, lazy exports, and import/runtime isolation from Torch. Torch BSA plus tensor-adapter/parity tests passed 55 with 23 skips, including both layouts, mixed-framework rejection by the JAX API, and Torch execution with JAX unavailable. API-index tool unit tests: 8 passed; the canonical-built-package runtime-index test skipped locally because its package-root variable was unset. git diff --check passed. This is not reported as a passing full runtime API inventory. Reused owner-started OSS pipeline 68289973 and verified its full GitHub head through mirror commit metadata. Style/Docs passed and the Rubin OSS lane passed; SM100 shards remained pending. The three red jobs retain the earlier unrelated signatures: the missing SdpaFwdDslSm100.template_params API-index entry, and SM80/90 gated-block imports requiring cutlass.experimental. Their current logs were inspected; this BSA namespace change does not touch those implementations.

Limitations: The pipeline is not fully green; pending SM100 CI is not counted as passing evidence. Local execution used a compatible existing native binding with the exact reviewed Python source, not a new native build. No fresh performance sweep for this final namespace-only diff. The earlier dispatch-overhead measurements apply to the superseded unified API, whose dispatch layer is now removed; they are not measurements of this head.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
Starting OSS CI for the new Torch namespace aliases at `132daad1b12f`; the previous run covered `49c506d65f01`. I will reuse the existing review evidence for unchanged implementations and check the new aliases separately.

Codex bot review — model `gpt-6-astra`.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `132daad1b12f` against base `55fdf97ca929`.

Re-reviewed the four-file delta from 49c506d65f01. The three cudnn.torch BSA exports are lazy aliases of the existing function objects; documentation and API inventory match. No new actionable finding. The earlier approval remains valid.

Validation: Exact-head SM100 run: 3 tests passed with zero skips, covering BHSD/BSHD JAX-to-Torch forward/backward parity, mixed-framework rejection, and a fresh process proving lazy alias identity plus Torch forward/backward execution with JAX unavailable. Confirmed the Torch/JAX BSA implementations, kernels, cache/stream handling and JAX bridge are unchanged from the previously reviewed head; reused the earlier validation for those implementations. git diff --check passed. Current-head Style, Docs and merge-requirements checks passed. Started OSS CI pipeline 68385414 and verified its mirror commit identifies this exact GitHub head. Full CI remains pending.

Limitations: The focused run used exact-head Python sources with a compatible existing native extension after verifying native-source identity. No fresh performance measurement: these aliases add no execution wrapper and retain the same function objects.

Comment only: this PR already has an active approval; preserving existing approvals.

**Comment by YangXu1990uiuc** [comment]
CI follow-up for `132daad1b12f`: Rubin OSS job 444040523 failed during Slurm/container setup (working-directory permission denied, then required plugin initialization failed), before pytest started. This is an infrastructure failure and provides no BSA test result. The SM100 lanes are still pending; the independent SM100 checks in the review passed.

Codex bot review — model `gpt-6-astra`.
